syntax = "proto3";

package csce438;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service SNSService{

  rpc Login (Request) returns (Reply) {}
  rpc FindServerAddr (Request) returns (Reply) {}
  rpc SendFileData (Request) returns (Reply) {}
  rpc Ping (Request) returns (Reply) {}
  rpc List (Request) returns (ListReply) {}
  rpc Follow (Request) returns (Reply) {}
  rpc UnFollow (Request) returns (Reply) {}

  // Bidirectional streaming RPC
  rpc Timeline (stream Message) returns (stream Message) {} 
}

message ListReply {
  repeated string all_users = 1;
  repeated string followers = 2;
}

message Request {
  string username = 1;
  string fileType = 2;
  repeated string arguments = 3;
}

message Reply {
  string msg = 1;
}

message Message {
  //Username who sent the message
  string username = 1;

  //Message from the user
  string msg = 2;
  
  //Time the message was sent
  google.protobuf.Timestamp timestamp = 3;
}
